omnigibson/simulator.py

```diff
@@ -603,26 +608,56 @@ def _on_contact(self, contact_headers, contact_data):
             headers = defaultdict(list)
             for contact_header in contact_headers:
                 actor0 = str(PhysicsSchemaTools.intToSdfPath(contact_header.actor0))
                 actor1 = str(PhysicsSchemaTools.intToSdfPath(contact_header.actor1))
                 # actor0/1 are prim paths for links that are in contact. Find the corresponding objects.
                 actor0_obj = self._scene.object_registry("prim_path", "/".join(actor0.split("/")[:-1]))
                 actor1_obj = self._scene.object_registry("prim_path", "/".join(actor1.split("/")[:-1]))
-                if actor0_obj is None or actor1_obj is None or not actor0_obj.initialized or not actor1_obj.initialized:
+                # If any of the objects cannot be found, skip
+                if actor0_obj is None or actor1_obj is None:
                     continue
-                headers[tuple(sorted((actor0_obj, actor1_obj), key=lambda x:x.uuid))].append(contact_header)
-
-            for (actor0_obj, actor1_obj) in combos:
+                # If any of the objects is not initialized, skip
+                if not actor0_obj.initialized or not actor1_obj.initialized:
+                    continue
+                # If any of the objects is not stateful, skip
                 if not isinstance(actor0_obj, StatefulObject) or not isinstance(actor1_obj, StatefulObject):
                     continue
+                # If any of the objects doesn't have states that require on_contact callbacks, skip
+                if len(actor0_obj.states.keys() & self.object_state_types_on_contact) == 0 or len(actor1_obj.states.keys() & self.object_state_types_on_contact) == 0:
+                    continue
+                headers[tuple(sorted((actor0_obj, actor1_obj), key=lambda x: x.name))].append(contact_header)
+
+            for (actor0_obj, actor1_obj) in headers:
                 for obj0, obj1 in [(actor0_obj, actor1_obj), (actor1_obj, actor0_obj)]:
                     for state_type in self.object_state_types_on_contact:
                         if state_type in obj0.states:
                             obj0.states[state_type].on_contact(obj1, headers[(actor0_obj, actor1_obj)], contact_data)
 
+    def _on_simulation_event(self, event):
+        """
+        This callback will be invoked if there is any simulation event. Currently it only processes JOINT_BREAK event.
+        """
+        if gm.ENABLE_OBJECT_STATES:
+            if event.type == int(SimulationEvent.JOINT_BREAK):
+                joint_path = str(PhysicsSchemaTools.decodeSdfPath(event.payload["jointPath"][0], event.payload["jointPath"][1]))
+                obj = None
+                # TODO: recursively try to find the parent object of this joint
+                tokens = joint_path.split("/")
+                for i in range(2, len(tokens) + 1):
+                    obj = self._scene.object_registry("prim_path", "/".join(tokens[:i]))
+                    if obj is not None:
+                        break
```

Comment:
Given a descendant joint, is there a good way to find the corresponding object in object_registry?

Reply:
mmmmm...not really, unless we want to register all the objects under every joint it owns also. I think your recursive way probably makes the most sense

Reply:
btw, did you see how slow this call was (how much overhead it adds) if we run it in a full scene like Rs_int?

Reply:
We could potentially register these mapping in `JointBreakSubscribedStateMixin` but it will make things stateful. Since the overhead here is very small (joint break event is very rare), it should be fine for now.